src/control/tools/StrokeHandler.cpp

```diff
@@ -58,24 +62,72 @@ auto StrokeHandler::onMotionNotifyEvent(const PositionInputData& pos) -> bool {
         return false;
     }
 
-    double zoom = xournal->getZoom();
-    double x = pos.x / zoom;
-    double y = pos.y / zoom;
-    int pointCount = stroke->getPointCount();
+    if (pos.pressure == 0) {
+        /**
+         * Some devices emit a move event with pressure 0 when lifting the stylus tip
+         * Ignore those events
+         */
+        return true;
+    }
 
-    Point currentPoint(x, y);
+    stabilizer->processEvent(pos);
+    return true;
+}
+
+void StrokeHandler::paintTo(const Point& point) {
+
+    int pointCount = stroke->getPointCount();
 
     if (pointCount > 0) {
-        if (!validMotion(currentPoint, stroke->getPoint(pointCount - 1))) {
-            return true;
+        Point endPoint = stroke->getPoint(pointCount - 1);
+        double distance = point.lineLengthTo(endPoint);
+        if (distance < PIXEL_MOTION_THRESHOLD) {  //(!validMotion(point, endPoint)) {
+            return;
+        }
+        if (point.z != Point::NO_PRESSURE && stroke->getToolType() == STROKE_TOOL_PEN) {
+            /**
+             * Both device and tool are pressure sensitive
+             */
+            if (endPoint.z != Point::NO_PRESSURE) {
+                /**
+                 * Avoid issues at the beginning of the stroke
+                 */
+                double widthDelta = (point.z - endPoint.z) * stroke->getWidth();
+                if (-widthDelta > MAX_WIDTH_VARIATION || widthDelta > MAX_WIDTH_VARIATION) {
+                    /**
+                     * If the width variation is to big, decompose into shorter segments.
+                     * Those segments can not be shorter than PIXEL_MOTION_THRESHOLD
+                     */
+                    int nbSteps = std::min(std::ceil(std::abs(widthDelta) / MAX_WIDTH_VARIATION),
+                                           std::floor(distance / PIXEL_MOTION_THRESHOLD));
+                    Point increment((point.x - endPoint.x) / (double)nbSteps, (point.y - endPoint.y) / (double)nbSteps,
+                                    widthDelta / (double)nbSteps);
+                    endPoint.z *= stroke->getWidth();
+                    endPoint.z += increment.z;
+                    stroke->setLastPressure(endPoint.z);
+
+#ifdef STAB_DEBUG
+                    printf("Add %d points after point %3d | %9.4f | %9.4f | %9.7f |\n", nbSteps - 1, pointCount,
+                           endPoint.x, endPoint.y, endPoint.z);
+#endif
+
+                    for (int i = 1; i < nbSteps; i++) {  // The last step is done below
+                        endPoint.x += increment.x;
+                        endPoint.y += increment.y;
+                        endPoint.z += increment.z;
+                        drawSegmentTo(endPoint);
```

Comment:
**Question:** Incrementing in a loop might be a bit faster than multipliing, but could introduce floating-point-epsilon errors. Dunno if this could result in visual bugs?

Reply:
This loop can only be a dozen iteration long at worse (for a very very thick stroke and a pressure value dropping from 1 to 0). I doubt there could be any visible rounding error with so few iterations